Fix aggregator_unavailable_apiservice gauge

dgrisonnet opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

**What type of PR is this?**

/kind bug

<!--
Add one of the following kinds:
/kind bug
/kind cleanup
/kind documentation
/kind feature
/kind design

Optionally add one or more of the following kinds if applicable:
/kind api-change
/kind deprecation
/kind failing-test
/kind flake
/kind regression
-->

**What this PR does / why we need it**:

When an apiservice is deleted, its relative `aggregator_unavailable_apiservice` metric remains with the value of the last availability observed. Hence, if an apiservice is deleted while being unavailable, the metric remains marked as unavailable.
This presents some issues when alerting on unavailable apiservices as deleted apiservices might trigger the alert indefinitely.

To solve this issue, the `aggregator_unavailable_apiservice` metric should only reflect the availability of existing apiservices.

This is achievable by using a custom Collector instead of a GaugeVec and create throw-away metrics based on an apiservice lister output. With this approach, on deletion, the apiservice will not be listed anymore, resulting in its availability metric not being exposed.

**Which issue(s) this PR fixes**:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #92671

**Special notes for your reviewer**:

The alert that I previously mentioned is the `AggregatedAPIDown` alert from the kubernetes-mixin project: https://github.com/kubernetes-monitoring/kubernetes-mixin/blob/fe19e65e64183f4a54a4584c792608851f68a1e9/alerts/kube_apiserver.libsonnet#L88-L101
In my opinion, this alert is correct and the problem lies into the `aggregator_unavailable_apiservice` metric which is why I opened this PR.

**Does this PR introduce a user-facing change?**:
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
Fixed a bug where aggregator_unavailable_apiservice metrics were reported for deleted apiservices.
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```
/sig instrumentation

dgrisonnet committed:
kube-aggregator: fix apiservice availability gauge

When an apiservice is deleted, its relative
aggregator_unavailable_apiservice metric remains with the value of the
last availability observed. Hence, if an apiservice is deleted while
being unavailable, the metric remains marked as unavailable.
This presents some problems when alerting on unavailable apiservices
as deleted apiservices might trigger the alert indefinitely.

To solve this issue, the aggregator_unavailable_apiservice metric should
only reflect the availability of existing apiservices.

This is achievable by using a custom Collector instead of a GaugeVec and
create throw-away metrics based on an apiservice lister output. With
this approach, on deletion, the apiservice will not be listed anymore,
resulting in its availability metric not being exposed.

Signed-off-by: Damien Grisonnet <dgrisonn@redhat.com>

dgrisonnet commented:
/assign @lavalamp

fedebongio commented:
/assign @deads2k 
/unassign @lavalamp 
/triage accepted
/cc @logicalhan

sttts commented:
/ok-to-test

sttts commented on staging/src/k8s.io/kube-aggregator/pkg/controllers/status/metrics.go:
```diff
@@ -38,17 +44,65 @@ var (
+// Register registers metrics for status controller.
+func Register(apiServiceLister listers.APIServiceLister) {
+	registerMetrics.Do(func() {
+		legacyregistry.CustomMustRegister(newAPIServiceStatusCollector(apiServiceLister))
```
while touching this, can we turn legacyregstry into a parameter of Register? Now it's half a global construct and the other half isn't.

sttts replied:
the `sync.Once` does not play well with this :-/ Maybe the `sync.Once` belongs where we call Register? These globals are ugly....

dgrisonnet replied:
legacyregistry is a package here and the global registry it uses isn't exported.

dgrisonnet replied:
I would agree with you that this isn't very clean, but I mostly followed what's done in other packages.

I don't think it is meaningful to move the `sync.Once`  to where we call Register. It is used to prevent registering the same metrics multiple times into the global registry. If such a thing occurs, it would result in a panic :/. To me, it should be very local to this Register function and thus not be moved elsewhere.
Also, we won't be able to turn it into a local variable as it would make it possible to register the metrics multiple times.

The correct approach here would be to get rid of the global registry, but this would require a bigger rewrite.

sttts commented on staging/src/k8s.io/kube-aggregator/pkg/controllers/status/metrics.go:
```diff
@@ -38,17 +44,65 @@ var (
-	legacyregistry.MustRegister(unavailableCounter)
-	legacyregistry.MustRegister(unavailableGauge)
+// CollectWithStability implements the metrics.StableCollector interface.
+func (c *apiServiceStatusCollector) CollectWithStability(ch chan<- metrics.Metric) {
```
when is this called?

dgrisonnet replied:
This is called when the apiserver metrics are scraped. The exact interval between each scrape depends on how it is configured, but it's usually every 30s.

sttts replied:
ack. It's what I expected.

deads2k commented:
/assign @sttts 
/unassign

sttts commented on staging/src/k8s.io/kube-aggregator/pkg/controllers/status/metrics.go:
```diff
@@ -38,17 +44,65 @@ var (
+// CollectWithStability implements the metrics.StableCollector interface.
+func (c *apiServiceStatusCollector) CollectWithStability(ch chan<- metrics.Metric) {
+	apiServiceList, _ := c.apiServiceLister.List(labels.Everything())
+	for _, apiService := range apiServiceList {
```
the more I think about this pattern, the more I dislike it. IMO it inverses direction in the wrong way: metrics are global (they shouldn't be, but it's where we are), and leaking a client into a global construct makes this wrongly shaped construct even bigger. Instead we should strive to remove globals and any logic connected to them.

dgrisonnet replied:
when I was writing my first draft for this PR, I took a different approach that is quite different from what's done in other packages but considering your comment, it might be better suited. The custom collector had a map, that was updated by the controller with the availabilities of the apiservices. Upon collection, the metrics were created based on the information stored in the map instead of the ones from the lister.

sttts commented:
How about adding a metrics call at https://github.com/kubernetes/kubernetes/blob/135de8e0a1050af6496d19aa4bb31635ef9d111e/staging/src/k8s.io/kube-aggregator/pkg/controllers/status/available_controller.go#L212 to set the value to zero?

sttts commented:
Actually not sure this will solve it. Is there a way to make the aggregator server struct the "owner" of a metric?

dgrisonnet commented:
> How about adding a metrics call to set the value to zero?

That's something that has also been suggested by @logicalhan in the issue, but I had some concerns about it: https://github.com/kubernetes/kubernetes/issues/92671#issuecomment-717799237

> Actually not sure this will solve it. Is there a way to make the aggregator server struct the "owner" of a metric?

Yes, we could wrap the Prometheus metrics in a Metrics structure owned by the controller. We could then have getters/setters like: `c.metrics.SetAPIServiceAvailable`.

sttts commented:
> Yes, we could wrap the Prometheus metrics in a Metrics structure owned by the controller. We could then have getters/setters like: `c.metrics.SetAPIServiceAvailable`.

That's probably much better. Thinking along the lines of https://github.com/kubernetes-sigs/metrics-server/pull/587. I really like that factoring where the metrics are not global.

dgrisonnet commented:
Ok, I will try this approach. Thank you for the review.

dgrisonnet commented:
@sttts I've updated this PR according to our previous discussions, can you please have another look at it?
I removed most of the global states, but we still need to keep a global sync.Once to avoid registering the metrics multiple times in the global registry. As you suggested, I moved this global closer to where we call the metric registration and it indeed looks way better. With this change, it should now be possible to register the metrics with multiple registry without being limited by the global registry constraints.
Also, the metrics are now owned by the controller, and clients are not leaked into the collection logic anymore. All the logic is now self-contained in the collector itself.

sttts commented on staging/src/k8s.io/kube-aggregator/pkg/controllers/status/metrics.go:
```diff
@@ -30,25 +31,125 @@ import (
+				apiServiceName,
+			)
+		} else {
+			ch <- metrics.NewLazyConstMetric(
```
nit: the call can be factored out. Only set value depending on isAvailable.

sttts commented on staging/src/k8s.io/kube-aggregator/pkg/controllers/status/available_controller.go:
```diff
@@ -52,6 +53,8 @@ import (
 	"k8s.io/kube-aggregator/pkg/controllers"
 )
 
+var registerMetrics sync.Once
```
```
// making sure we only register metrics once into legacy registry
var registerIntoLegacyRegistry sync.Once
```

sttts commented on staging/src/k8s.io/kube-aggregator/pkg/controllers/status/available_controller.go:
```diff
@@ -93,6 +96,8 @@ type AvailableConditionController struct {
 	// by the controller only vary on the dynamic cert/key.
 	tlsCache *tlsTransportCache
+
+	metrics *availabilityMetrics
```
// metrics registered into legacy registry

sttts commented on staging/src/k8s.io/kube-aggregator/pkg/controllers/status/available_controller.go:
```diff
@@ -203,12 +209,21 @@ func NewAvailableConditionController(
 
 	c.syncFn = c.sync
 
+	var err error
```
// TODO: decouple from legacyregistry

sttts commented:
/lgtm
/approve

ialidzhikov commented:
@dgrisonnet , does it makes sense to also cherry-pick this PR into release-1.17, release-1.18 and release-.1.19?

dgrisonnet commented:
@ialidzhikov, I am not too familiar with k/k backport process but I don't think it makes sense here. Even though the bug exists in past releases, it's a corner case with a fairly low impact so I don't think it is worth cherry-picking.

sttts commented:
This needs backports.
